src/execution/aggregate_hashtable.cpp

```diff
@@ -378,33 +378,40 @@ void GroupedAggregateHashTable::FetchAggregates(DataChunk &groups, DataChunk &re
 	for (idx_t aggr_idx = 0; aggr_idx < aggregates.size(); aggr_idx++) {
 		D_ASSERT(result.ColumnCount() > aggr_idx);
 
-		VectorOperations::Gather::Set(addresses, result.data[aggr_idx], groups.size());
+		auto &target = result.data[aggr_idx];
+		auto &aggr = aggregates[aggr_idx];
+		aggr.function.finalize(addresses, aggr.bind_data, target, result.size());
+		VectorOperations::AddInPlace(addresses, aggr.payload_size, result.size());
 	}
 }
 
 template <class T>
 static void TemplatedScatter(VectorData &gdata, Vector &addresses, const SelectionVector &sel, idx_t count,
-                             idx_t type_size) {
+                             idx_t col_offset, idx_t col_idx) {
 	auto data = (T *)gdata.data;
-	auto pointers = FlatVector::GetData<uintptr_t>(addresses);
+	auto pointers = FlatVector::GetData<data_ptr_t>(addresses);
+
 	if (!gdata.validity.AllValid()) {
 		for (idx_t i = 0; i < count; i++) {
 			auto pointer_idx = sel.get_index(i);
 			auto group_idx = gdata.sel->get_index(pointer_idx);
-			auto ptr = (T *)pointers[pointer_idx];
-
-			T store_value = !gdata.validity.RowIsValid(group_idx) ? NullValue<T>() : data[group_idx];
-			Store<T>(store_value, (data_ptr_t)ptr);
-			pointers[pointer_idx] += type_size;
+			auto ptr = pointers[pointer_idx] + col_offset;
+
+			auto isnull = !gdata.validity.RowIsValid(group_idx);
+			T store_value = isnull ? NullValue<T>() : data[group_idx];
```

Comment:
`NullValue<T>` is not necessary anymore, right? Perhaps just store what is in `data[group_idx]`, or add an `#ifdef DEBUG` around using it?

Reply:
I was *very* nervous about changing this because it would make binary images non-deterministic, which in turn would make testing annoying. Can we revisit `NullValue` and its uses in a later PR? I still feel that `NullValue` is useful, it is just what it returns that we should consider.

Reply:
I think that is fine. I'm not entirely sure in what use cases you would directly inspect the contents of a hash table dump such that this would be useful, but I imagine the cost of doing this is also negligible. Perhaps something to benchmark?

Reply:
`NullValue` is used in other places. In particular, I was worried about `Vector::Serialize`, which seems to be used in odd places (like `BLOB`s?)

Reply:
Indeed, for sure NullValue needs to stay until that code is also rewritten. I was thinking specifically about the use of NullValue here.